pandas/index.pyx

```diff
@@ -426,6 +426,67 @@ cdef class Int64Engine(IndexEngine):
 
         return result
 
+cdef class UInt64Engine(IndexEngine):
+
+    cdef _get_index_values(self):
+        return algos.ensure_uint64(self.vgetter())
+
+    cdef _make_hash_table(self, n):
+        return _hash.UInt64HashTable(n)
+
+    def _call_monotonic(self, values):
+        return algos.is_monotonic_uint64(values, timelike=False)
+
+    def get_pad_indexer(self, other, limit=None):
+        return algos.pad_uint64(self._get_index_values(), other,
+                                limit=limit)
+
+    def get_backfill_indexer(self, other, limit=None):
+        return algos.backfill_uint64(self._get_index_values(), other,
+                                     limit=limit)
+
+    cdef _check_type(self, object val):
+        hash(val)
+        if util.is_bool_object(val):
+            raise KeyError(val)
+        elif util.is_float_object(val):
+            raise KeyError(val)
+
+    cdef _maybe_get_bool_indexer(self, object val):
+        cdef:
```

Comment:
you can prob just use fused types for routines like this (for the scalar & array type)